src/lib/backoff.ts

```diff
@@ -0,0 +1,88 @@
+import { promisify } from 'util';
+
+const sleep = promisify(setTimeout);
+
+export type BackoffCallback = {
+	retryCount: number;
+	delay: number;
+	error: any;
+};
+
+export type Options = {
+	attempt: number;
+	maxDelay: number;
+	minDelay: number;
+	maxRetries: number;
+	retryCallback?: (info: BackoffCallback) => void;
+};
+
+const DEFAULT_OPTIONS: Partial<Options> = {
+	attempt: 0,
+	maxDelay: 15000,
+	minDelay: 15000,
+	maxRetries: Infinity,
+};
+
+/**
+ * Retries a function with exponential delay between calls.
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ *
+ * Supports exceptions which return `retryAfter` value to specify the backoff duration.
+ * Will call `retryCallback` if passed in options as a way for callers to react to retries
+ */
+export function withBackoff<T extends (...args: any[]) => any>(
+	func: T,
+	options: Partial<Options> = {},
+) {
+	// If the function returns a promise, unwrap the promise return type
+	// otherwise use the actual return
+	type TReturn = ReturnType<T> extends Promise<infer R> ? R : ReturnType<T>;
+
+	const normalizedOptions: Options = {
+		...DEFAULT_OPTIONS,
+		...options,
+	} as Options;
+
+	return async function wrapped(...args: Parameters<T>): Promise<TReturn> {
+		let error: any;
+		try {
+			return await func(...args);
+		} catch (e) {
+			error = e;
+			if (e.retryAfter) {
```

Comment:
Might want to check that `retryAfter` is of the correct type.

You could do something like

```
export class RetryError extends Error {
    constructor(public statusMessage?: string, public retryAfter: number) {
        super(statusMessage);
    }
}

// Type guard
const isRetryError = (x: unknown): x is RetryError => x != null && x instanceof Error && Number.isInteger(x.retryAfter);
```

And then inside the if, you can use the type guard
```
if (isRetryError(e)) {
    // e.retryAfter will be available here
}
```

Reply:
I actually found that there is an existing StatusError type in errors module which has a retryAfter parameter!